File: poses_viewer/src/drill2d/drillMetrics.ts
/**
 * Per-rep extraction of the in-plane metrics at the stroke's wrist-speed peak
 * — 1:1 mirror of Kotlin DrillMetrics. Score-gated per joint, sanity-bounded per
 * value, then MEDIAN over a ±70 ms window (keypoints are unsmoothed — one junk
 * frame must not shift the rep value).
 */
import { Handedness, PoseFrame2D, StrokeCycle2D } from './types'
import { DEFAULT_MIN_SCORE } from './facing'
import { elbowAngle, hipFlexion, kneeBend, shoulderAngle, shoulderTilt, torsoLean } from './angles2d'
import { isSane } from './sanityBounds'

export const METRIC = {
  ELBOW_ANGLE: 'elbow_angle',
  SHOULDER_ANGLE: 'shoulder_angle',
  KNEE_BEND: 'knee_bend',
  TORSO_LEAN: 'torso_lean',
  SHOULDER_TILT: 'shoulder_tilt',
  HIP_FLEXION: 'hip_flexion',
} as const

export const ALL_KEYS: string[] = [
  METRIC.ELBOW_ANGLE, METRIC.SHOULDER_ANGLE, METRIC.KNEE_BEND,
  METRIC.TORSO_LEAN, METRIC.SHOULDER_TILT, METRIC.HIP_FLEXION,
]

/** Half-width of the peak-metric smoothing window, in ms (±2 frames at 30 fps). */
export const DEFAULT_PEAK_RADIUS_MS = 70

/**
 * Extract the in-plane metrics for a single frame.
 * Each metric is score-gated by the angle function (returns null when below minScore),
 * then sanity-bounded — matching Kotlin's collect-then-filter approach.
 */
export function extractAtFrame(
  frame: PoseFrame2D, handedness: Handedness, xScale: number, minScore = DEFAULT_MIN_SCORE,
): Record<string, number> {
  const kp = frame.keypoints
  if (kp.length === 0) return {}
  const raw: Record<string, number | null> = {
    [METRIC.ELBOW_ANGLE]: elbowAngle(kp, handedness, xScale, minScore),
    [METRIC.SHOULDER_ANGLE]: shoulderAngle(kp, handedness, xScale, minScore),
    [METRIC.KNEE_BEND]: kneeBend(kp, handedness, xScale, minScore),
    [METRIC.TORSO_LEAN]: torsoLean(kp, xScale, minScore),
    [METRIC.SHOULDER_TILT]: shoulderTilt(kp, xScale, minScore),
    [METRIC.HIP_FLEXION]: hipFlexion(kp, handedness, xScale, minScore),
  }
  const out: Record<string, number> = {}
  for (const [key, value] of Object.entries(raw)) {
    if (value !== null && isSane(key, value)) out[key] = value
  }
  return out
}

function medianOf(values: number[]): number {
  const sorted = [...values].sort((a, b) => a - b)
  const mid = Math.floor(sorted.length / 2)
  return sorted.length % 2 === 1 ? sorted[mid] : (sorted[mid - 1] + sorted[mid]) / 2
}

/**
 * Per-rep metrics: MEDIAN of each metric over the frames within ±radiusMs of peakFrame.
 * Radius is computed as Math.trunc(radiusMs / intervalMs) — integer division matching Kotlin.
 * Window is inclusive [peakFrame - radius, peakFrame + radius], clamped to array bounds.
 * Each frame is independently score-gated and sanity-bounded via extractAtFrame.
 * At coarse intervals (radius < interval) this degrades to the single peak frame.
 */
export function extractAtPeak(
  frames: PoseFrame2D[],
  peakFrame: number,
  handedness: Handedness,
  xScale: number,
  intervalMs: number,
  minScore = DEFAULT_MIN_SCORE,
  radiusMs = DEFAULT_PEAK_RADIUS_MS,
): Record<string, number> {
  if (intervalMs <= 0) throw new Error(`intervalMs must be > 0, got ${intervalMs}`)
  if (peakFrame < 0 || peakFrame >= frames.length) {
    throw new Error(`peakFrame ${peakFrame} out of bounds for ${frames.length} frames`)
  }
  const radius = Math.trunc(radiusMs / intervalMs)
  const lo = Math.max(peakFrame - radius, 0)
  const hi = Math.min(peakFrame + radius, frames.length - 1)
  const byKey: Record<string, number[]> = {}
  for (let i = lo; i <= hi; i++) {
    for (const [key, value] of Object.entries(extractAtFrame(frames[i], handedness, xScale, minScore))) {
      ;(byKey[key] ??= []).push(value)
    }
  }
  const out: Record<string, number> = {}
  for (const [key, values] of Object.entries(byKey)) out[key] = medianOf(values)
  return out
}

// ---------------------------------------------------------------------------
// Per-phase extraction
// ---------------------------------------------------------------------------

/** The three stroke phases at which metrics are measured. */
export type Phase = 'backswing' | 'contact' | 'followthrough'

/**
 * Declares which phases each metric should be measured at.
 * Exported so table UI and reference-range modules can import it without
 * duplicating the mapping.
 *
 * This is a CURATED SUBSET of all metrics — not every metric gets per-phase
 * treatment. The `satisfies Partial<...>` check enforces that any key present
 * is a valid MetricKey (typo'd keys are compile-time errors), while still
 * allowing the subset to omit metrics.
 *
 * Movement-bracketing rule:
 *   - Arm metrics (elbow, shoulder): backswing + followthrough (the swing arc).
 *     Contact is the noisiest instant for both — RTMPose mislocates the fast forearm
 *     and the shoulder sweeps through its full range, so no stable ideal exists at contact.
 *   - Legs/trunk metrics (knee, hip, torso): backswing + contact (load → drive).
 *     Followthrough is excluded — it is a recovery phase that rotates out of the camera
 *     plane, so 2D coaching values are unreliable there (no ideal to grade against).
 *
 * shoulder_tilt is intentionally excluded from METRIC_PHASES — it remains a
 * single-instant colored cell (using standard.ranges directly).
 */
export const METRIC_PHASES = {
  // Legs/trunk — load → drive; followthrough excluded (rotation-corrupted)
  [METRIC.KNEE_BEND]:       ['backswing', 'contact'],
  [METRIC.HIP_FLEXION]:     ['backswing', 'contact'],
  [METRIC.TORSO_LEAN]:      ['backswing', 'contact'],
  // Arm — swing arc (замах → завершення); contact excluded (noisiest instant)
  [METRIC.ELBOW_ANGLE]:     ['backswing', 'followthrough'],
  [METRIC.SHOULDER_ANGLE]:  ['backswing', 'followthrough'],
} satisfies Partial<Record<(typeof METRIC)[keyof typeof METRIC], Phase[]>>

/**
 * Extract per-metric, per-phase values for a full stroke cycle.
 *
 * Phase anchor frames:
 *   backswing   → cycle.drive.startFrame  (the turn; only emitted when cycle.backswing != null)
 *   contact     → cycle.drive.peakFrame   (same anchor as extractAtPeak today)
 *   followthrough → cycle.drive.endFrame
 *
 * Each phase is computed via the existing extractAtPeak (±radiusMs median window).
 * extractAtPeak is called AT MOST ONCE per distinct anchor frame — results are
 * memoized by frame index so metrics sharing the same anchor reuse the window.
 *
 * A phase value is `number | null` (null when extractAtPeak gated the metric at
 * that frame). The `backswing` key is omitted entirely from a metric's record
 * when cycle.backswing is null.
 */
export function extractPerPhase(
  cycle: StrokeCycle2D,
  frames: PoseFrame2D[],
  handedness: Handedness,
  xScale: number,
  intervalMs: number,
  minScore = DEFAULT_MIN_SCORE,
  radiusMs = DEFAULT_PEAK_RADIUS_MS,
): Record<string, Partial<Record<Phase, number | null>>> {
  // Map phase → anchor frame index (omit backswing when unpaired).
  const phaseAnchors: Partial<Record<Phase, number>> = {
    contact:       cycle.drive.peakFrame,
    followthrough: cycle.drive.endFrame,
  }
  if (cycle.backswing !== null) {
    phaseAnchors.backswing = cycle.drive.startFrame
  }

  // Pre-flight: validate each anchor frame that will actually be used.
  for (const [phase, frameIndex] of Object.entries(phaseAnchors) as [Phase, number][]) {
    if (frameIndex < 0 || frameIndex >= frames.length) {
      throw new Error(
        `extractPerPhase: ${phase} anchor frame ${frameIndex} out of bounds for ${frames.length} frames`,
      )
    }
  }

  // Memoize extractAtPeak calls by frame index so shared anchors are not
  // recomputed (e.g. two metrics at 'contact' → same window, one call).
  const cache = new Map<number, Record<string, number>>()
  function peakAt(frameIndex: number): Record<string, number> {
    let cached = cache.get(frameIndex)
    if (cached === undefined) {
      cached = extractAtPeak(frames, frameIndex, handedness, xScale, intervalMs, minScore, radiusMs)
      cache.set(frameIndex, cached)
    }
    return cached
  }

  const result: Record<string, Partial<Record<Phase, number | null>>> = {}

  for (const [metricKey, phases] of Object.entries(METRIC_PHASES)) {
    const phaseRecord: Partial<Record<Phase, number | null>> = {}
    for (const phase of phases) {
      const anchorFrame = phaseAnchors[phase]
      if (anchorFrame === undefined) {
        // backswing phase requested but cycle is unpaired — omit entirely.
        continue
      }
      const extracted = peakAt(anchorFrame)
      // Keep null explicitly when extractAtPeak gated this metric out.
      phaseRecord[phase] = extracted[metricKey] !== undefined ? extracted[metricKey] : null
    }
    result[metricKey] = phaseRecord
  }

  return result
}

File: shared/src/commonMain/kotlin/com/ttcoachai/shared/drill/DrillMetrics.kt
package com.ttcoachai.shared.drill

import com.ttcoachai.shared.analysis.AngleCalculations2D
import com.ttcoachai.shared.models.Handedness
import com.ttcoachai.shared.models.PoseFrame2D

/**
 * Per-frame extraction of the five Phase 2 in-plane metrics (context doc §3) at the
 * stroke's wrist-speed peak. Score-gated per joint; sanity-bounded per value.
 *
 * Thin wrapper: the generalized algorithm — any [MetricSpec] list, not just these
 * five — lives in [MovementMetrics]. Kept for API compatibility; behavior for the
 * five core metrics is unchanged bit-for-bit.
 */
object DrillMetrics {

    const val METRIC_ELBOW_ANGLE = "elbow_angle"
    const val METRIC_SHOULDER_ANGLE = "shoulder_angle"
    /**
     * hip-knee-ankle angle at the stroke's wrist-speed peak (strike), NOT the backswing.
     * [extractAtPeak] only ever computes one value per rep from the frames around
     * [peakFrame] — there is no separate backswing-phase extraction — so an editor-authored
     * "knees · backswing" target has no metric to bind to; only "knees · strike" is wireable
     * via a [com.ttcoachai.shared.analysis.BaselineRule.RangeRule] against this key.
     */
    const val METRIC_KNEE_BEND = "knee_bend"
    const val METRIC_TORSO_LEAN = "torso_lean"
    /**
     * Still defined and used by [com.ttcoachai.shared.analysis.AngleCalculations2D.shoulderTilt],
     * [VoicePresetCatalog], and [CoreMessageTemplates] — just no longer part of [PEAK_KEYS]/[ALL_KEYS]
     * (removed from [CoreMetricSpecs.ALL], the peak-extracted spec list).
     */
    const val METRIC_SHOULDER_TILT = "shoulder_tilt"
    /** Legacy 3D owns `"follow_through_angle"` ([MetricKeyDisjointTest]) — this 2D key takes the `_2d` suffix. */
    const val METRIC_FOLLOW_THROUGH_ANGLE_2D = "follow_through_angle_2d"
    const val METRIC_STROKE_SPEED = "stroke_speed"
    const val METRIC_COIL_RATIO = "coil_ratio"

    /** The four metrics extracted at the stroke's wrist-speed peak (see [CoreMetricSpecs.ALL]). */
    val PEAK_KEYS = listOf(
        METRIC_ELBOW_ANGLE, METRIC_SHOULDER_ANGLE, METRIC_KNEE_BEND, METRIC_TORSO_LEAN
    )

    /** The three derived metrics computed from the peak metrics + stroke window (later tasks own the computation). */
    val DERIVED_KEYS = listOf(
        METRIC_FOLLOW_THROUGH_ANGLE_2D, METRIC_STROKE_SPEED, METRIC_COIL_RATIO
    )

    val ALL_KEYS = PEAK_KEYS + DERIVED_KEYS

    /** Half-width of the peak-metric smoothing window, in ms (±2 frames at 30 fps). */
    const val DEFAULT_PEAK_RADIUS_MS = 70L

    fun extractAtFrame(
        frame: PoseFrame2D,
        handedness: Handedness,
        xScale: Float,
        minScore: Float = AngleCalculations2D.DEFAULT_MIN_SCORE
    ): Map<String, Double> =
        MovementMetrics.extractAtFrame(frame, handedness, xScale, CoreMetricSpecs.ALL, minScore)

    /**
     * Per-rep metrics: MEDIAN of each metric over the frames within ±[radiusMs] of
     * [peakFrame] (DESIGN_LIMITATIONS L-05 — keypoints are unsmoothed, so a single
     * raw frame feeds RTMPose jitter straight into the baseline). Median, not mean:
     * one junk frame inside the window must not shift the rep's value. Each frame
     * is score-gated and sanity-bounded independently via [extractAtFrame]; at
     * coarse intervals (radius < interval) this degrades to the single peak frame.
     */
    fun extractAtPeak(
        frames: List<PoseFrame2D>,
        peakFrame: Int,
        handedness: Handedness,
        xScale: Float,
        intervalMs: Long,
        minScore: Float = AngleCalculations2D.DEFAULT_MIN_SCORE,
        radiusMs: Long = DEFAULT_PEAK_RADIUS_MS
    ): Map<String, Double> =
        MovementMetrics.extractAtPeak(frames, peakFrame, handedness, xScale, intervalMs, CoreMetricSpecs.ALL, minScore, radiusMs)
}
